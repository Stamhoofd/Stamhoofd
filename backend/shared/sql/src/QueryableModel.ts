import { Model, SQLResultNamespacedRow } from '@simonbackx/simple-database';
import { SQLSelect } from './SQLSelect';
import { SQL } from './SQL';
import { SQLDelete } from './SQLDelete';
import { SQLUpdate } from './SQLUpdate';
import { SQLInsert } from './SQLInsert';

export class QueryableModel extends Model {
    static select<T extends typeof Model>(this: T): SQLSelect<InstanceType<T>> {
        const transformer = (row: SQLResultNamespacedRow): InstanceType<T> => {
            const d = (this as T).fromRow(row[this.table] as any) as InstanceType<T> | undefined;

            if (!d) {
                throw new Error('EmailTemplate not found');
            }

            return d;
        };

        const select = new SQLSelect(transformer, SQL.wildcard());
        return select.from(SQL.table(this.table));
    }

    static delete(): SQLDelete {
        return SQL.delete().from(SQL.table(this.table));
    }

    static update(): SQLUpdate {
        return SQL.update(SQL.table(this.table));
    }

    static insert(): SQLInsert {
        return SQL.insert(SQL.table(this.table));
    }
}
