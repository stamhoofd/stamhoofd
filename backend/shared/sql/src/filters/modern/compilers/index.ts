export * from './less';
export * from './greater';
export * from './equals';
export * from './in';
