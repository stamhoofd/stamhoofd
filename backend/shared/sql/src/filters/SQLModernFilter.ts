import { SimpleError } from '@simonbackx/simple-errors';
import { compileFilter, FilterCompiler, FilterCompilerSelector, FilterDefinitions, StamhoofdFilter } from '@stamhoofd/structures';
import { SQLExpression, SQLExpressionOptions, SQLQuery } from '../SQLExpression';
import { SQLJsonUnquote, SQLJsonValue } from '../SQLJsonExpressions';
import { SQLWhere, SQLWhereAnd, SQLWhereNot, SQLWhereOr } from '../SQLWhere';
import { $equalsSQLFilterCompiler, $greaterThanSQLFilterCompiler, $inSQLFilterCompiler, $lessThanSQLFilterCompiler } from './compilers';
import { isJSONColumn } from './helpers/isJSONColumn';

export type SQLSyncFilterRunner = (column: SQLCurrentColumn) => SQLWhere;
export type SQLFilterRunner = (column: SQLCurrentColumn) => Promise<SQLWhere> | SQLWhere;
export type SQLFilterCompiler = FilterCompiler<SQLFilterRunner>;
export type SQLFilterDefinitions = FilterDefinitions<SQLFilterRunner>;
export type SQLFilterCompilerSelector = FilterCompilerSelector<SQLFilterRunner>;

export enum SQLValueType {
    /** At the root of a select */
    Table = 'Table',

    /** Column with type string */
    String = 'String',

    /** MySQL Datetime */
    Datetime = 'Datetime',

    /** Column with type number */
    Number = 'Number',

    /** Column with type boolean, meaning 1 or 0 */
    Boolean = 'Boolean',

    /** True or false in JSON */
    JSONBoolean = 'JSONBoolean',
    JSONString = 'JSONString',

    JSONNumber = 'JSONNumber',

    /** [...] */
    JSONArray = 'JSONArray',

    /** {...} */
    JSONObject = 'JSONObject',
}

export type SQLCurrentColumn = {
    expression: SQLExpression;

    /**
     * MySQL nullable. Please fill this in correctly! If a value can be null, can not exist (=mysql null), or can be JSONNull, set this to true.
     *
     * Mainly > and < operators will make sure the behaviour is consistent with MySQL sorting (normally comparing with null will always return false in MySQL)
     */
    nullable?: boolean;

    /**
     * JSON nullable
     */

    /**
     * Type of this column, use to normalize values received from filters
     */
    type: SQLValueType;
    checkPermission?: () => Promise<void>;
};

export function createColumnFilter(column: SQLCurrentColumn): SQLFilterCompiler {
    return (filter: StamhoofdFilter, filters: SQLFilterCompilerSelector) => {
        const runner = $andSQLFilterCompiler(filter, filters);

        return (_: SQLCurrentColumn) => {
            return runner({
                nullable: false,
                ...column,
            });
        };
    };
}

function filterDefinitionsToSelector(definitions: SQLFilterDefinitions): SQLFilterCompilerSelector {
    return (key: string) => {
        return definitions[key];
    };
}

export function $andSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompilerSelector): SQLFilterRunner {
    const runners = compileSQLFilter(filter, filters);

    return async (column: SQLCurrentColumn) => {
        const wheres = (await Promise.all(
            runners.map(runner => (runner(column))),
        ));

        return new SQLWhereAnd(wheres);
    };
}

export function $orSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompilerSelector): SQLFilterRunner {
    const runners = compileSQLFilter(filter, filters);

    return async (column: SQLCurrentColumn) => {
        const wheres = (await Promise.all(
            runners.map(runner => (runner(column))),
        ));

        return new SQLWhereOr(wheres);
    };
}

export function $notSQLFilterCompiler(filter: StamhoofdFilter, filters: SQLFilterCompilerSelector): SQLFilterRunner {
    const andRunner = $andSQLFilterCompiler(filter, filters);

    return async (column: SQLCurrentColumn) => {
        return new SQLWhereNot(await andRunner(column));
    };
}

function invertFilterCompiler(compiler: SQLFilterCompiler): SQLFilterCompiler {
    return (filter: StamhoofdFilter, compilers: SQLFilterCompilerSelector) => {
        const runner = compiler(filter, compilers);
        return async (column) => {
            return new SQLWhereNot(await runner(column));
        };
    };
}

export const baseSQLFilterCompilers: SQLFilterDefinitions = {
    $and: $andSQLFilterCompiler,
    $or: $orSQLFilterCompiler,
    $not: $notSQLFilterCompiler,
    $eq: $equalsSQLFilterCompiler,
    $neq: invertFilterCompiler($equalsSQLFilterCompiler),

    $lt: $lessThanSQLFilterCompiler,
    $gt: $greaterThanSQLFilterCompiler,
    $lte: invertFilterCompiler($greaterThanSQLFilterCompiler),
    $gte: invertFilterCompiler($lessThanSQLFilterCompiler),

    $in: $inSQLFilterCompiler,
};

const compileSQLFilter = compileFilter<SQLFilterRunner>;

export const SQLRootExpression: SQLExpression = {
    getSQL(options?: SQLExpressionOptions): SQLQuery {
        throw new SimpleError({
            code: 'invalid_filter',
            message: 'Root level filters are not allowed to use $eq or $neq',
        });
    },
};

export async function compileToSQLFilter(filter: StamhoofdFilter, filters: SQLFilterDefinitions): Promise<SQLWhere> {
    if (filter === null) {
        return new SQLWhereAnd([]); // No filter, return empty where
    }
    const runner = $andSQLFilterCompiler(filter, filterDefinitionsToSelector(filters));
    return await runner({
        expression: SQLRootExpression,
        type: SQLValueType.Table,
        nullable: false,
    });
};

/**
 * Casts json strings, numbers and booleans to native MySQL types. This includes json null to mysql null.
 */
export function normalizeColumn(column: SQLCurrentColumn): SQLCurrentColumn {
    if (column.type === SQLValueType.JSONString) {
        return {
            expression: new SQLJsonValue(column.expression, 'CHAR'),
            type: SQLValueType.String,
        };
    }

    if (column.type === SQLValueType.JSONBoolean) {
        return {
            expression: new SQLJsonValue(column.expression, 'UNSIGNED'),
            type: SQLValueType.Boolean,
        };
    }

    if (column.type === SQLValueType.JSONNumber) {
        return {
            expression: new SQLJsonValue(column.expression, 'UNSIGNED'),
            type: SQLValueType.Number,
        };
    }
    return column;
}
