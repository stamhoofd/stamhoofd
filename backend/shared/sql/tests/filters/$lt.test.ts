import { baseModernSQLFilterCompilers, createColumnFilter, SQLValueType } from '../../src/filters/modern/SQLModernFilter';
import { SQL } from '../../src/SQL';
import { TableDefinition, test, testMatch } from '../utils';

describe('$lt', () => {
    /**
     * Tests that should be repeated for all filter types
     */
    describe('Common checks', () => {
        // todo
    });

    it('can filter on $lt', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $lt: 18,
                },
            },
            filters,
            query: {
                query: '`default`.`age` < ?',
                params: [18],
            },
        });
    });

    it('can invert $lt inside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                age: {
                    $not: {
                        $lt: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` >= ?',
                params: [18],
            },
        });
    });

    it('can invert $lt outside', async () => {
        const filters = {
            ...baseModernSQLFilterCompilers,
            age: createColumnFilter({ expression: SQL.column('age'), type: SQLValueType.Number, nullable: false }),
        };

        await test({
            filter: {
                $not: {
                    age: {
                        $lt: 18,
                    },
                },
            },
            filters,
            query: {
                query: '`default`.`age` >= ?',
                params: [18],
            },
        });
    });

    describe('MySQL Behaviour', () => {
        const tableDefinition: TableDefinition = {
            settings: {
                type: 'json',
                nullable: true,
            },
            name: {
                type: 'varchar',
                nullable: true,
            },
            age: {
                type: 'int',
                nullable: true,
            },
        };
        const filters = {
            ...baseModernSQLFilterCompilers,
            'name': createColumnFilter(
                { expression: SQL.column('name'), type: SQLValueType.String, nullable: true },
            ),
            'age': createColumnFilter(
                { expression: SQL.column('age'), type: SQLValueType.Number, nullable: true },
            ),
            'settings.name': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.name'), type: SQLValueType.JSONString, nullable: true },
            ),
            'settings': createColumnFilter({ expression: SQL.column('settings'), type: SQLValueType.JSONObject, nullable: true }),
            'settings.enabled': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.enabled'), type: SQLValueType.JSONBoolean, nullable: true },
            ),
            'settings.age': createColumnFilter({ expression: SQL.jsonValue(SQL.column('settings'), '$.age'), type: SQLValueType.JSONNumber, nullable: true },
            ),
        };

        it('Comparing string values is case insensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: 'alice',
                    },
                    {
                        name: 'Alice',
                    },
                    {
                        name: 'ALICE',
                    },
                ],
                doMatch: [
                    {
                        name: {
                            $lt: 'b',
                        },
                    },
                    {
                        name: {
                            $lt: 'alicea',
                        },
                    },
                    {
                        name: {
                            $lt: 'Alicea',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        name: {
                            $lt: 'Alice',
                        },
                    },
                    {
                        name: {
                            $lt: 'aLice',
                        },
                    },
                    {
                        name: {
                            $lt: 'ALICE',
                        },
                    },
                    {
                        name: {
                            $lt: 'alice',
                        },
                    },
                    {
                        name: {
                            $lt: 'a',
                        },
                    },
                ],
            });
        });

        it('Comparing JSON string values is case sensitive', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: 'alice',
                        },
                    },
                    {
                        settings: {
                            name: 'Alice',
                        },
                    },
                    {
                        settings: {
                            name: 'ALICE',
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $lt: 'alicea',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: 'Alicea',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: 'ALICEA',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.name': {
                            $lt: 'Alice',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: 'aLice',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: 'ALICE',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: 'alice',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: '',
                        },
                    },
                ],
            });
        });

        it('Comparing numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        age: 10,
                    },
                ],
                doNotMatch: [
                    {
                        age: {
                            $lt: 10,
                        },
                    },
                    {
                        age: {
                            $lt: 9,
                        },
                    },
                    {
                        age: {
                            $lt: 0,
                        },
                    },
                    {
                        age: {
                            $lt: -9,
                        },
                    },
                    {
                        age: {
                            $lt: 1,
                        },
                    },
                ],
                doMatch: [
                    {
                        age: {
                            $lt: 12,
                        },
                    },
                    {
                        age: {
                            $lt: 20,
                        },
                    },
                ],
            });
        });

        it('Comparing JSON numbers values is numberic', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: 10,
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.age': {
                            $lt: 10,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: 9,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: 0,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: -9,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: 1,
                        },
                    },
                ],
                doMatch: [
                    {
                        'settings.age': {
                            $lt: 12,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: 20,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON numbers', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            age: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doMatch: [
                    {
                        'settings.age': {
                            $lt: -1000,
                        },
                    },
                    {
                        'settings.age': {
                            $lt: 0,
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.age': {
                            $lt: null,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for JSON strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        settings: {
                            name: null,
                        },
                    },
                    {
                        settings: null,
                    },
                    {
                        settings: {},
                    },
                ],
                doMatch: [
                    {
                        'settings.name': {
                            $lt: 'aaaa',
                        },
                    },
                    {
                        'settings.name': {
                            $lt: '',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        'settings.name': {
                            $lt: null,
                        },
                    },
                ],
            });
        });

        it('NULL values are always the smallest for strings', async () => {
            await testMatch({
                tableDefinition,
                filters,
                rows: [
                    {
                        name: null,
                    },
                ],
                doMatch: [
                    {
                        name: {
                            $lt: '',
                        },
                    },
                    {
                        name: {
                            $lt: 'aaaa',
                        },
                    },
                ],
                doNotMatch: [
                    {
                        name: {
                            $lt: null,
                        },
                    },
                ],
            });
        });
    });
});
