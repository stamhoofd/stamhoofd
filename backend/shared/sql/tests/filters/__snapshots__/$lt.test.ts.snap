// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lt":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for filter: {"settings.age":{"$lt":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lt":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lt":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lt":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON numbers values is numberic: SQL Query for not match filter: {"settings.age":{"$lt":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lt":"ALICEA"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lt":"Alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for filter: {"settings.name":{"$lt":"alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lt":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lt":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lt":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lt":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing JSON string values is case sensitive: SQL Query for not match filter: {"settings.name":{"$lt":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lt":12}} 1`] = `
{
  "params": [
    12,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for filter: {"age":{"$lt":20}} 1`] = `
{
  "params": [
    20,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lt":-9}} 1`] = `
{
  "params": [
    -9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lt":1}} 1`] = `
{
  "params": [
    1,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lt":9}} 1`] = `
{
  "params": [
    9,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing numbers values is numberic: SQL Query for not match filter: {"age":{"$lt":10}} 1`] = `
{
  "params": [
    10,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`age\` IS NULL OR \`test_table\`.\`age\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lt":"Alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lt":"alicea"}} 1`] = `
{
  "params": [
    "alicea",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for filter: {"name":{"$lt":"b"}} 1`] = `
{
  "params": [
    "b",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lt":"ALICE"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lt":"Alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lt":"a"}} 1`] = `
{
  "params": [
    "a",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lt":"aLice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour Comparing string values is case insensitive: SQL Query for not match filter: {"name":{"$lt":"alice"}} 1`] = `
{
  "params": [
    "alice",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lt":-1000}} 1`] = `
{
  "params": [
    -1000,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for filter: {"settings.age":{"$lt":0}} 1`] = `
{
  "params": [
    0,
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") IS NULL) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.age") < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON numbers: SQL Query for not match filter: {"settings.age":{"$lt":null}} 1`] = `"always false"`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for filter: {"settings.name":{"$lt":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for filter: {"settings.name":{"$lt":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE (JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") = CAST('null' AS JSON) OR JSON_EXTRACT(\`test_table\`.\`settings\`,"$.name") IS NULL) OR JSON_VALUE(\`test_table\`.\`settings\`,"$.name" RETURNING CHAR CHARACTER SET utf8mb4) < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for JSON strings: SQL Query for not match filter: {"settings.name":{"$lt":null}} 1`] = `"always false"`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for filter: {"name":{"$lt":""}} 1`] = `
{
  "params": [
    "",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for filter: {"name":{"$lt":"aaaa"}} 1`] = `
{
  "params": [
    "aaaa",
  ],
  "query": "SELECT \`test_table\`.* FROM \`test_table\` WHERE \`test_table\`.\`name\` IS NULL OR \`test_table\`.\`name\` < ?",
}
`;

exports[`$lt MySQL Behaviour NULL values are always the smallest for strings: SQL Query for not match filter: {"name":{"$lt":null}} 1`] = `"always false"`;
